Source/WebCore/loader/FrameLoader.cpp

```diff
@@ -1219,6 +1219,35 @@ static NavigationNavigationType determineNavigationType(FrameLoadType loadType,
     return NavigationNavigationType::Push;
 }
 
+// https://html.spec.whatwg.org/multipage/browsing-the-web.html#url-and-history-update-steps
+void FrameLoader::updateURLAndHistory(const URL& newURL, RefPtr<SerializedScriptValue>&& stateObject, NavigationHistoryBehavior historyHandling)
+{
+    RefPtr document = m_frame->document();
+    RefPtr documentLoader = protectedDocumentLoader();
+    CheckedRef history = m_frame->checkedHistory();
+
+    if (documentLoader->isInitialAboutBlank())
+        historyHandling = NavigationHistoryBehavior::Replace;
+
+    // FIXME: Validate that stateObject deserializes without exception.
+
+    if (!stateObject)
+        stateObject = history->currentItem()->stateObject();
+
+    document->setURL(newURL);
+    setOutgoingReferrer(newURL);
+    documentLoader->replaceRequestURLForSameDocumentNavigation(newURL);
+
+    if (historyHandling == NavigationHistoryBehavior::Replace)
+        history->replaceState(WTFMove(stateObject), newURL.string());
+    else
+        history->pushState(WTFMove(stateObject), newURL.string());
+
+    // Consider the navigation completed.
```

Comment:
Why are these two lines needed? AFAICS they are not listed as part of the spec algorithm.

Reply:
This isn't the ideal place for it. I wanted to ensure the navigation was marked as completed but I'll keep that to a different change.